spring-cloud-data-shell/src/main/java/org/springframework/cloud/data/shell/command/ModuleCommands.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.data.shell.command;
+
+import java.util.List;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.cloud.data.core.ModuleType;
+import org.springframework.cloud.data.rest.client.ModuleOperations;
+import org.springframework.cloud.data.rest.resource.DetailedModuleRegistrationResource;
+import org.springframework.cloud.data.rest.resource.DetailedModuleRegistrationResource.Option;
+import org.springframework.cloud.data.rest.resource.ModuleRegistrationResource;
+import org.springframework.cloud.data.shell.config.CloudDataShell;
+import org.springframework.hateoas.PagedResources;
+import org.springframework.shell.core.CommandMarker;
+import org.springframework.shell.core.annotation.CliAvailabilityIndicator;
+import org.springframework.shell.core.annotation.CliCommand;
+import org.springframework.shell.core.annotation.CliOption;
+import org.springframework.shell.support.table.TableRow;
+import org.springframework.stereotype.Component;
+import org.springframework.shell.support.table.Table;
+import org.springframework.shell.support.table.TableHeader;
+
+
+/**
+ * Commands for working with modules. Allows retrieval of information about
+ * available modules, as well as creating and removing module registrations.
+ *
+ * @author Glenn Renfro
+ * @author Eric Bottard
+ * @author Florent Biville
+ * @author David Turanski
+ * @author Patrick Peralta
+ */
+@Component
+public class ModuleCommands implements CommandMarker {
+
+	private final static String LIST_MODULES = "module list";
+
+	private final static String MODULE_INFO = "module info";
+
+	private final static String UNREGISTER_MODULE = "module unregister";
+
+	private static final String REGISTER_MODULE = "module register";
+
+	@Autowired
+	private CloudDataShell cloudDataShell;
+
+
+	@CliAvailabilityIndicator({LIST_MODULES, MODULE_INFO, UNREGISTER_MODULE, REGISTER_MODULE})
+	public boolean available() {
+		return cloudDataShell.getCloudDataOperations() != null;
+	}
+
+	@CliCommand(value = MODULE_INFO, help = "Get information about a module")
+	public String moduleInfo(
+			@CliOption(mandatory = true,
```

Comment:
Ah, I see.
So completion would work with that single "name" arg, but then why require
a --type arg?

What can't work is this: `register --type --name <TAB>` because the
completer for the name arg doesn't "see" the other args, namely --type here

On Mon, Aug 31, 2015 at 5:33 PM, Patrick Peralta <notifications@github.com>
wrote:

> In
> spring-cloud-data-shell/src/main/java/org/springframework/cloud/data/shell/command/ModuleCommands.java
> <https://github.com/spring-cloud/spring-cloud-data/pull/68#discussion_r38324028>
> :
>
> > +	private final static String UNREGISTER_MODULE = "module unregister";
> > +
> > +	private static final String REGISTER_MODULE = "module register";
> > +
> > +	@Autowired
> > +	private CloudDataShell cloudDataShell;
> > +
> > +
> > +	@CliAvailabilityIndicator({LIST_MODULES, MODULE_INFO, UNREGISTER_MODULE, REGISTER_MODULE})
> > +	public boolean available() {
> > +		return cloudDataShell.getCloudDataOperations() != null;
> > +	}
> > +
> > +	@CliCommand(value = MODULE_INFO, help = "Get information about a module")
> > +	public String moduleInfo(
> > +			@CliOption(mandatory = true,
>
> I added this method to handle the :...is there a way to make this work
> with autocompletion?
>
> /** * Return a {@link QualifiedModuleName} for the given arguments. * If {@code type} is {@code null}, the module type may be obtained * from the module name if the module name is in the format * {@code name:type}. * * @param name module name * @param type module type; may be {@code null} * @return {@code QualifiedModuleName} for the provided arguments */private QualifiedModuleName processArgs(String name, ModuleType type) {
>     if (type == null) {
>         String[] split = name.split("\\:");
>         if (split.length != 2) {
>             throw new IllegalArgumentException(
>                     String.format("Expected format of 'name:type' for module name %s", name));
>         }
>         return new QualifiedModuleName(split[0], ModuleType.valueOf(split[1]));
>     }
>     else {
>         return new QualifiedModuleName(name, type);
>     }
> }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spring-cloud/spring-cloud-data/pull/68/files#r38324028>
> .
>